### ISSUE TITLE
Random crashes when accessing m_fxChannels[] field

### ISSUE THREAD

**Issue by Bloody95** [LMMS/lmms#2043]
I just experienced frequent crashes on not so complex compositions (4 or 5 instruments, no VST).
I recompiled this morning the end of branch (sha1 388dfbc38ccf4a5e2779b2260e5a4d0dbb1c332a) on Linux and had crashes when simply playing the tracks.

I run gdb on lmms and had most of the crashes when accessing m_fxChannels, in FxMixer::masterMix() (when clearing the audio buffer) or in FxMixerView::updateFaders().

I am totally new to the lmms code (just cloned it and compiled it for the 1st time today), so the fix I did is probably not correct and may impact performances, but the following is working so far for me : 
1) I took the lock (and released it of course) in updateFaders() in the same way it is taken in masterMix(),
2) instead of releasing the lock in masterMix() in the middle of the function, I expanded its range up to the end of the function.

```
diff --git a/src/core/FxMixer.cpp b/src/core/FxMixer.cpp
index b99acc3..87bddf4 100644
--- a/src/core/FxMixer.cpp
+++ b/src/core/FxMixer.cpp
@@ -508,22 +508,22 @@ void FxMixer::masterMix( sampleFrame * _buf )
                {
                        MixerWorkerThread::startAndWaitForJobs();
                }
-               m_sendsMutex.unlock();
-       }
-       
-       const float v = m_fxChannels[0]->m_volumeModel.value();
-       MixHelpers::addSanitizedMultiplied( _buf, m_fxChannels[0]->m_buffer, v, fpp );
 
-       // clear all channel buffers and
-       // reset channel process state
-       for( int i = 0; i < numChannels(); ++i)
-       {
-               engine::mixer()->clearAudioBuffer( m_fxChannels[i]->m_buffer, engine::mixer()->framesPerPeriod() );
-               m_fxChannels[i]->reset();
-               m_fxChannels[i]->m_queued = false;
-               // also reset hasInput
-               m_fxChannels[i]->m_hasInput = false;
-               m_fxChannels[i]->m_dependenciesMet = 0;
+               const float v = m_fxChannels[0]->m_volumeModel.value();
+               MixHelpers::addSanitizedMultiplied( _buf, m_fxChannels[0]->m_buffer, v, fpp );
+
+               // clear all channel buffers and
+               // reset channel process state
+               for( int i = 0; i < numChannels(); ++i)
+               {
+                       engine::mixer()->clearAudioBuffer( m_fxChannels[i]->m_buffer, engine::mixer()->framesPerPeriod() );
+                       m_fxChannels[i]->reset();
+                       m_fxChannels[i]->m_queued = false;
+                       // also reset hasInput
+                       m_fxChannels[i]->m_hasInput = false;
+                       m_fxChannels[i]->m_dependenciesMet = 0;
+               }
+               m_sendsMutex.unlock();
        }
 }
```
and : 
```
diff --git a/src/gui/FxMixerView.cpp b/src/gui/FxMixerView.cpp
index f750bbc..4bad101 100644
--- a/src/gui/FxMixerView.cpp
+++ b/src/gui/FxMixerView.cpp
@@ -482,33 +482,37 @@ void FxMixerView::updateFaders()
        FxMixer * m = engine::fxMixer();
 
        // apply master gain
-       m->m_fxChannels[0]->m_peakLeft *= engine::mixer()->masterGain();
-       m->m_fxChannels[0]->m_peakRight *= engine::mixer()->masterGain();
-
-       for( int i = 0; i < m_fxChannelViews.size(); ++i )
+       if( m->m_sendsMutex.tryLock() )
        {
-               const float opl = m_fxChannelViews[i]->m_fader->getPeak_L();
-               const float opr = m_fxChannelViews[i]->m_fader->getPeak_R();
-               const float fall_off = 1.2;
-               if( m->m_fxChannels[i]->m_peakLeft > opl )
-               {
-                       m_fxChannelViews[i]->m_fader->setPeak_L( m->m_fxChannels[i]->m_peakLeft );
-                       m->m_fxChannels[i]->m_peakLeft = 0;
-               }
-               else
-               {
-                       m_fxChannelViews[i]->m_fader->setPeak_L( opl/fall_off );
-               }
+               m->m_fxChannels[0]->m_peakLeft *= engine::mixer()->masterGain();
+               m->m_fxChannels[0]->m_peakRight *= engine::mixer()->masterGain();
 
-               if( m->m_fxChannels[i]->m_peakRight > opr )
-               {
-                       m_fxChannelViews[i]->m_fader->setPeak_R( m->m_fxChannels[i]->m_peakRight );
-                       m->m_fxChannels[i]->m_peakRight = 0;
-               }
-               else
+               for( int i = 0; i < m_fxChannelViews.size(); ++i )
                {
-                       m_fxChannelViews[i]->m_fader->setPeak_R( opr/fall_off );
+                       const float opl = m_fxChannelViews[i]->m_fader->getPeak_L();
+                       const float opr = m_fxChannelViews[i]->m_fader->getPeak_R();
+                       const float fall_off = 1.2;
+                       if( m->m_fxChannels[i]->m_peakLeft > opl )
+                       {
+                               m_fxChannelViews[i]->m_fader->setPeak_L( m->m_fxChannels[i]->m_peakLeft );
+                               m->m_fxChannels[i]->m_peakLeft = 0;
+                       }
+                       else
+                       {
+                               m_fxChannelViews[i]->m_fader->setPeak_L( opl/fall_off );
+                       }
+
+                       if( m->m_fxChannels[i]->m_peakRight > opr )
+                       {
+                               m_fxChannelViews[i]->m_fader->setPeak_R( m->m_fxChannels[i]->m_peakRight );
+                               m->m_fxChannels[i]->m_peakRight = 0;
+                       }
+                       else
+                       {
+                               m_fxChannelViews[i]->m_fader->setPeak_R( opr/fall_off );
+                       }
                }
+               m->m_sendsMutex.unlock();
        }
 }
```

**Comment by Umcaruje** [comment]
@Bloody95 can you still reproduce this? @Fastigium has been fixed a lot of issues concerning the Mixer Channels, so there is a big chance this was fixed too.

**Comment by Bloody95** [comment]
Selon Umcaruje <notifications@github.com>:

> @Bloody95 can you still reproduce this? @Fastigium has been fixed a lot of
> issues concerning the Mixer Channels, so there is a big chance this was fixed
> too.
>
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/LMMS/lmms/issues/2043#issuecomment-203634477


I will check it as soon as I can.

Regards

**Comment by zonkmachine** [comment]
The FxMixer crashes was largely fixed in 1.2.0-RC1 and the last gllitches has been ironed out in current master. Closing. Please reopen if you see still see these crashes.

@Bloody95 When you submit code to lmms (which I encourage you to do!) you should do it as a [Pull Request](https://github.com/LMMS/lmms/wiki/Submitting-a-patch). If you submit it like this it's easy to skip over it as a possible back trace.
